nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/adls/PutADLSFile.java

```diff
@@ -0,0 +1,267 @@
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import static org.apache.nifi.processors.adls.ADLSConstants.*;
+
+@InputRequirement(InputRequirement.Requirement.INPUT_REQUIRED)
+@Tags({"hadoop", "ADLS", "put", "copy", "filesystem", "restricted"})
+@CapabilityDescription("Write FlowFile data to Azure Data Lake Store (ADLS)")
+@ReadsAttributes({
+        @ReadsAttribute(attribute = "filename", description = "The name of the file written to ADLS comes from the value of this attribute."),
+        @ReadsAttribute(attribute = "filepath", description = "The relative path to the file on ADLS comes from the value of this attribute.")
+})
+@WritesAttributes({
+        @WritesAttribute(attribute = "filename", description = "The name of the file written to ADLS is stored in this attribute."),
+        @WritesAttribute(attribute = "absolute.adls.path", description = "The absolute path to the file on ADLS is stored in this attribute.")
+})
+@Restricted("Provides operator the ability to write to any file that NiFi has access to in ADLS.")
+public class PutADLSFile extends ADLSAbstractProcessor {
+
+    private static final String PART_FILE_EXTENSION = ".nifipart";
+    private static final String REPLACE_RESOLUTION = "replace";
+    private static final String FAIL_RESOLUTION = "fail";
+    private static final String APPEND_RESOLUTION = "append";
+
+    protected static final AllowableValue REPLACE_RESOLUTION_AV = new AllowableValue(REPLACE_RESOLUTION,
+            REPLACE_RESOLUTION, "Replaces the existing file if any.");
+    protected static final AllowableValue FAIL_RESOLUTION_AV = new AllowableValue(FAIL_RESOLUTION, FAIL_RESOLUTION,
+            "Penalizes the flow file and routes it to failure.");
+    protected static final AllowableValue APPEND_RESOLUTION_AV = new AllowableValue(APPEND_RESOLUTION, APPEND_RESOLUTION,
+            "Appends to the existing file if any, creates a new file otherwise.");
+
+    // properties
+    protected static final PropertyDescriptor CONFLICT_RESOLUTION = new PropertyDescriptor.Builder()
+            .name("Conflict Resolution Strategy")
+            .description("Indicates what should happen when a file " +
+                    "with the same name already exists in the output directory")
+            .required(true)
+            .defaultValue(FAIL_RESOLUTION_AV.getValue())
+            .allowableValues(REPLACE_RESOLUTION_AV, FAIL_RESOLUTION_AV, APPEND_RESOLUTION_AV)
+            .build();
+
+    protected static final PropertyDescriptor DIRECTORY = new PropertyDescriptor.Builder()
+            .name("Directory")
+            .description("The directory where files will be written")
+            .required(true)
+            .addValidator(StandardValidators.ATTRIBUTE_EXPRESSION_LANGUAGE_VALIDATOR)
+            .expressionLanguageSupported(true)
+            .build();
+
+    protected static final PropertyDescriptor UMASK = new PropertyDescriptor.Builder()
```

Comment:
default value? required? validator? is it OK if the value is null?

Reply:
default value is decided from server if null is passed, custom validator already being used, null is OK.